### ISSUE TITLE
push packagecloud versions for linuxmint/{rebecca,rafaela,rosa}

### ISSUE THREAD

**Issue by javabrett** [github/git-lfs#1074]
This is a request for releng to push packages to packagecloud for at least `linuxmint/rebecca`, `linuxmint/rafaela` and `linuxmint/rosa`, the current linuxmint LTS releases.  Those are also the push-string specs required.  I imagine that these packages would be identical to their `ubuntu/trusty` upstream equivalents.

The current `git-lfs/script.deb.sh` will create repo specs such as:

```
deb https://packagecloud.io/github/git-lfs/linuxmint/ rebecca main
deb-src https://packagecloud.io/github/git-lfs/linuxmint/ rebecca main
```

Currently, these have to be changed to:

```
deb https://packagecloud.io/github/git-lfs/ubuntu/ trusty main
deb-src https://packagecloud.io/github/git-lfs/ubuntu/ trusty main
```

... in order for installation to be able to proceed.

linuxmint remains the most popular distribution, so it makes sense to push debs for it too.

**Comment by andyneff** [comment]
AFAIK the same Ubuntu packages should work on Mint.

**Comment by javabrett** [comment]
@andyneff Correct.

**Comment by javabrett** [comment]
I'm going to open a conversation with packagecloud about how the same package might be able to be shared across distro and perhaps distro-versions without duplicating the uploads, and still being able to use their nice install repo-setup script.

**Comment by technoweenie** [comment]
@javabrett That sounds great. :metal:

**Comment by javabrett** [comment]
Just to be clear, the motivation here is: packagecloud host the installer package repos, and provide neat scripts to help those on Linux who are less familiar with their (apt, rpm/Yum) package-manager and repo URLs.  These scripts are `script.deb.sh`, `script.rpm.sh` etc.

The problem is that these resolve specifically the installer's exact distribution and version (scripts call these `os` and `dist` respectively, somewhat confusingly to me anyway), so on say LinuxMint, these will automatically resolve to e.g. `os=linuxmint`, `dist=rebecca`.  If we don't upload such a package, the script succeeds, creates an invalid `deb` URL as above, then the install fails.  Presumably the same applies for ol ~= el,, centos ~= el etc.

I contacted packagecloud.  If we don't want to upload duplicate packages for compatible releases (per my original proposal, now seeming less-good to me), then the best workaround at the moment is to either not use the script and add the deb manually, then manually do the remaining required steps such as installing the signing-key etc., or if you want to use the script, you can pass it a compatible (upstream) distribution-name (as `os`), e.g. on LinuxMint/Rebecca `curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | os=ubuntu dist=trusty sudo -E bash`.

I guess for now we could upload each package once only, named for the upstream-most distribution+version it is compatible-with, then name in install-documentation all of the compatible downstream distributions/versions that should use that as an override, and provide the above command.  There might be exceptions - for example, even if for a release debian==downstream-ubuntu, Ubuntu might be considered popular/important-enough to create a proper upload for.  LinuxMint is pretty popular too though.

So to whoever is cutting/uploading the next release: do you have the map of number of package-versions, and which distribution-version combinations that are compatible, or suspected-compatible with each?  That would then help everyone on e.g. LinuxMint to get the command right and pick-up either the Ubuntu or Debian repos.

If that makes sense and with the information above, I can propose install-doc changes and close #1076.

**Comment by technoweenie** [comment]
> do you have the map of number of package-versions, and which distribution-version combinations that are compatible, or suspected-compatible with each?

It's tracked here: 

https://github.com/github/git-lfs/blob/9c7155dc780b47b704d21309c12eceef797fc893/script/packagecloud.rb#L21

I don't mind uploading the same packages multiple times. But, it sounds like people can install the PackageCloud repo by running the deb script with hard coded os/dist values. Once installed, future package updates through apt or yum will just work for Fedora, Linux Mint, and Ubuntu, even though the original packages are uploaded to PackageCloud once, for Enterprise Linux and Debian.

That sounds good to me :+1:

**Comment by javabrett** [comment]
> It's tracked here:
>
>https://github.com/github/git-lfs/blob/9c7155dc780b47b704d21309c12eceef797fc893/script/packagecloud.rb#L21

Yep, I've got that map, and so originally proposed the additions to it with https://github.com/github/git-lfs/pull/1076/files#diff-b75c3a11a4c6f4186a87ff378b668101R41 .  The current maps supposes that Ubuntu is common-enough that it is worth the duplicated upload, 'fair nuff.

The current rough-edge is that if you assume that many folks installing on Linux arrive at https://git-lfs.github.com/ , see the enormous packagecloud install button offering a package-based install (which most folks will prefer as you say for the reason of auto-upgrades etc,), they arrive at https://packagecloud.io/github/git-lfs/install a fairly plain page but which offers the helpful `curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | sudo bash` for quick-install.

There's a couple of rough-edges with that:

1. It doesn't actually install `git-lfs`, it only installs the repo.  You have to follow-up with `sudo apt-get install git-lfs` to install the LFS client.  The packagecloud script alerts you to this with `The repository is setup! You can now install packages.`, but not every person will immediately realise that they haven't in-fact installed git-lfs.
1. Unless you are on one of the exact dists/releases (os/dist) mentioned in the upload-map, or you twig to the fact that you are on LinuxMint and read the script and realise that you are going to end-up with a bad deb line unless you run something like `curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | os=ubuntu dist=trusty sudo -E bash`, adding the `os=ubuntu dist=trusty ... -E`, you will end-up with an invalid repository.  Doesn't feel like a lot of folks who haven't used packagecloud before will twig to this first-go.

I guess what I'm saying is that given the sparsity of the instructions on the packagecloud install page (and assuming that doesn't change), to improve the install experience and likelihood of success, we might want to place a gh-page in-between highlighting some of these subtleties.

**Comment by technoweenie** [comment]
A separate page sounds great. Would you be able to submit a md document for the git-lfs repo? I can port that to git-lfs.github.com too. I have an internal issue to open source the gh-pages repo, but I probably won't do that before LFS v1.2 ships.

**Comment by technoweenie** [comment]
I've been thinking about this, and _almost_ asked another user what they thought. But, I keep coming back to the fact that we're offloading a bit of extra work on the user just because we don't want to upload the same package multiple times.

Better docs is definitely good, though. Even if we upload dup packages, describing how to customize the packagecloud scripts for unmapped distros could help people figure out a workaround.

**Comment by javabrett** [comment]
Yeah I have flip-flopped on uploading duplicates a few times. That task originally prompted me to chase-down the packagecloud ruby gem issue preventing same.

Accepting that bandwidth and storage are unlikely to be an issue at the current size of Git LFS artifacts, the only downside is coverage. There will always be another fringe distro that could use a compatible package but can't because of how the packagecloud helper-script process detects it.  Of course that user is entitled to work out the compatible upstream and update their apt list to reference that deb repo URL.

If we covered all the LTS versions of the top few rpm and apt packaged distros, very few users would need to refer to the "if you are not running RHEL (releases), Fedora (releases), Debian (releases), Ubuntu (Releases) or LinuxMint (Releases), please read _this_ to learn how to install a compatible package".  It's a matter of settling on the reasonable list of compatible duplicates and updating that `packagecloud.rb` map.

**Comment by javabrett** [comment]
See #1130 for a page I roughed-out.

During this process I was reminded of the consequences of fan-out on distro/release and any subsequent failure to publish those repos or dropping maintenance for them.

When I originally corrected my deb line `/linuxmint/ rebecca main` -> `/ubuntu/ trusty main`, I then installed 1.1.1.  I find that I still have that version as my packaged-install, because when 1.1.2 was pushed to packagecloud it only pushed a few releases, either intentionally or unnoticed as a consequence of #1102 or #1103.  There was no ubuntu repo pushed to packagecloud for 1.1.2, only debian made it, so I've not been offered an upgrade since.

Fot that reason, https://github.com/javabrett/git-lfs/blob/installing-doc/INSTALLING.md currently mentions only debian and not ubuntu.
